File: _61_RotateList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next or not k:
            return head#in case there's only one node/null, or k equals 0
        
        def list_len(a_head):
            n = 0
            if not a_head:
                pass
            else:
                while a_head:
                    a_head = a_head.next
                    n += 1
            return n
        
        def rotate_once(head_node):
            if not head_node:
                return head_node#null node
            new_head = None
            original_head = head_node
            while head_node.next.next:
                head_node = head_node.next
            new_head = head_node.next
            head_node.next = None
            new_head.next = original_head
            return new_head
        
        length = list_len(head)
        if length != 0:
            rotation = k % length
        else:
            rotation = k
        
        for _ in range(rotation):
            head = rotate_once(head)
            
        return head
        

File: _61_RotateList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return head;
        int len = 0;
        ListNode node = head;
        while (node != null) {
            node = node.next;
            len++;
        }
        k = k % len;
        if (k == 0) return head;
        node = head;
        int move = len - k;
        ListNode prev = null;
        while (move > 0) {
            prev = node;
            node = node.next;
            move--;
        }
        ListNode newHead = node;
        prev.next = null;
        // Check for node == null is covered by k == 0
        while (node.next != null) {
            node = node.next;
        }
        node.next = head;
        return newHead;
    }
}